### ISSUE TITLE
POST payload is being lost on Cookie based authentication

### ISSUE THREAD

**Issue by ojecborec** [zmartzone/mod_auth_openidc#443]
I’ve noticed interesting behavior when sending POST requests to OIDC protected resources.

1. Providing Access Token as Authorization Bearer header.

`curl -k -H "Authorization: Bearer $access_token" -X POST "https://localhost/api/hello " -H "Content-Type: application/json" -d '{"name":"Bob"}'`

Backend service receives JSON payload as expected.

2. Providing Access Token as cookie.

`curl -k -b "access_token=$access_token" -X POST "https://localhost/api/hello " -H "Content-Type: application/json" -d '{"name":"Bob"}'`

Payload is being lost. However when I comment out the **OIDCOAuthAcceptTokenAs post** line in my configuration file (see bellow) the payload is being received by the backend service. Note that I don't have to comment out the "OIDCOAuthAcceptTokenAs post" line when providing Access Token as Authorization Bearer header.

*Apache configuration*

```apacheconf 
OIDCProviderIssuer https://*****
OIDCProviderAuthorizationEndpoint https://*****

OIDCClientID my_ac_client
OIDCClientSecret myopenidentitysecret

OIDCRedirectURI /some_url
OIDCCryptoPassphrase mysecret

OIDCOAuthClientID "."
OIDCOAuthClientSecret "."
OIDCOAuthIntrospectionEndpoint https://*****
OIDCOAuthIntrospectionEndpointAuth bearer_access_token
OIDCOAuthIntrospectionClientAuthBearerToken ""
OIDCOAuthIntrospectionEndpointMethod GET
OIDCOAuthAccessTokenBindingPolicy disabled
OIDCOAuthTokenExpiryClaim expires_in relative optional
OIDCOAuthAcceptTokenAs header
OIDCOAuthAcceptTokenAs cookie:access_token
OIDCOAuthAcceptTokenAs post
OIDCOAuthAcceptTokenAs query

ProxyPass /api balancer://api_cluster
ProxyPassReverse /api balancer://api_cluster

<Proxy balancer://api_cluster>
    BalancerMember https://localhost:8443/api route=node1
</Proxy>

<Location /api>
    AuthType oauth20
    Require valid-user
    LogLevel debug
</Location>
```

*Backend service (Spring Boot)*

```java
@PostMapping("/api/hello")
public String hello(@RequestBody Map map) {
    return format("Hello %s from API!", map.getOrDefault("name", "Unknown"));
}
```

**Comment by zandbelt** [comment]
It seems that `OIDCOAuthAcceptTokenAs post` is broken - and actually never really worked... - and the order of code execution makes it so that it does not trigger for "header" but does trigger for "cookie"...

**Comment by zandbelt** [comment]
it would be great if you can testdrive: https://mod-auth-openidc.org/download/mod_auth_openidc-2.4.0rc19-1.el7.x86_64.rpm

**Comment by jhrozek** [comment]
@ojecborec since you are hitting the issue with RHEL/CentOS packages, feel free to also open a bug at bugzilla.redhat.com (and if you are a RHEL customer, a customer case) so that we can bring the patch over to RHEL..

**Comment by ojecborec** [comment]
Version 2.4.0rc19 seems to be sending payload but the encoded one.

1. As Authorization Bearer header.

`curl -k -H "Authorization: Bearer $access_token" -X POST "https://localhost/api/hello " -H "Content-Type: application/json" -d '{"name":"Bob"}'`

```json
{"name":"Bob"}'
```

2. As a cookie.

`curl -k -b "access_token=$access_token" -X POST "https://localhost/api/hello " -H "Content-Type: application/json" -d '{"name":"Bob"}'`

```json
%7B%22name%22%3A%22Bob%22%7D
```

**Comment by zandbelt** [comment]
that is fixed in 2.4.0rc20 https://mod-auth-openidc.org/download/mod_auth_openidc-2.4.0rc20-1.el7.x86_64.rpm
